src/server/scripts/Argus/AntorusTheBurningThrone/boss_garothi_worldbreaker.cpp

```diff
@@ -0,0 +1,884 @@
+        instance->SendBossKillCredit(ENCOUNTER_ID_GAROTHI_WORLDBREAKER);
+        instance->SendEncounterUnit(ENCOUNTER_FRAME_DISENGAGE, me);
+    }
+
+    void SpellHit(Unit* /*caster*/, SpellInfo const* spell) override
+    {
+        switch (spell->Id)
+        {
+            case SPELL_CANNON_CHOOSER:
+            {
+                Creature* decimator = instance->GetCreature(DATA_DECIMATOR);
+                Creature* annihilator = instance->GetCreature(DATA_ANNIHILATOR);
+
+                if ((_lastCanonEntry == NPC_ANNIHILATOR && decimator) || (decimator && !annihilator))
+                {
+                    decimator->CastSpell(decimator, SPELL_DECIMATION_SELECTOR, true);
+                    Talk(SAY_DECIMATION, decimator);
+                    _lastCanonEntry = NPC_DECIMATOR;
+                }
+                else if ((_lastCanonEntry == NPC_DECIMATOR && annihilator) || (annihilator && !decimator))
+                {
+                    uint8 count = std::max<uint8>(MIN_TARGETS_SIZE, std::ceil(me->GetMap()->GetPlayersCountExceptGMs() / 5));
+                    for (uint8 i = 0; i < count; i++)
+                    {
+                        float x = AnnihilationCenterReferencePos.GetPositionX() + cos(frand(0.0f, float(M_PI * 2))) * frand(15.0f, 30.0f);
+                        float y = AnnihilationCenterReferencePos.GetPositionY() + sin(frand(0.0f, float(M_PI * 2))) * frand(15.0f, 30.0f);
+                        float z = me->GetMap()->GetHeight(me->GetPhaseShift(), x, y, AnnihilationCenterReferencePos.GetPositionZ());
+                        annihilator->CastSpell(x, y, z, SPELL_ANNIHILATION_SUMMON, true);
+                    }
+
+                    annihilator->CastSpell(annihilator, SPELL_ANNIHILATION_DUMMY);
+                    annihilator->CastSpell(annihilator, SPELL_ANNIHILATION_SELECTOR);
+                    Talk(SAY_ANNIHILATION);
+                    _lastCanonEntry = NPC_ANNIHILATOR;
+                }
+                break;
+            }
+            default:
+                break;
+        }
+    }
+
+    void SpellHitTarget(Unit* target, SpellInfo const* spell) override
+    {
+        switch (spell->Id)
+        {
+            case SPELL_FEL_BOMBARDMENT_SELECTOR:
+                Talk(SAY_ANNOUNCE_FEL_BOMBARDMENT, target);
+                break;
+            default:
+                break;
+        }
+    }
+
+    void OnSuccessfulSpellCast(SpellInfo const* spell) override
+    {
+        switch (spell->Id)
+        {
+            case SPELL_APOCALYPSE_DRIVE_FINAL_DAMAGE:
+                if (_apocalypseDriveCount == 1)
+                    events.Reset();
+                events.ScheduleEvent(EVENT_REENGAGE_PLAYERS, 3s + 500ms);
+                HideCannons();
+                me->RemoveUnitFlag(UNIT_FLAG_NOT_SELECTABLE);
+                break;
+            default:
+                break;
+        }
+    }
+
+    void DamageTaken(Unit* /*attacker*/, uint32& damage) override
```

Comment:
Wouldn't it be better to just initialize two private variables that is initialized to 65/60 and 35/20 depending on difficulty and using those, instead of duplicating logic and checking difficulty every time damage is taken?

Reply:
Well, it's just a getter and a switch here so I don't see any major reason to just dump the numbers in advance unless you want me to do micro-optimizations

Reply:
Difficulty switching inside is not a thing for instances that have normal/heroic/mythic modes, it only applies to 10/25/10h/25h ones

Reply:
Allright. Please stall the pr for a little longer then. I'm currently working on Mythic difficulty mechanics as well which will take a little longer because I'm also getting rid of some hook usages in the meantime because I got complaints that SpellHit and SpellHitTarget should no longer be used. I personally would disagree because they are great for AI based actions but well.